crates/ruff_python_formatter/src/expression/mod.rs

```diff
@@ -271,6 +272,12 @@ impl<'ast> IntoFormat<PyFormatContext<'ast>> for Expr {
 ///
 /// This mimics Black's [`_maybe_split_omitting_optional_parens`](https://github.com/psf/black/blob/d1248ca9beaf0ba526d265f4108836d89cf551b7/src/black/linegen.py#L746-L820)
 fn can_omit_optional_parentheses(expr: &Expr, context: &PyFormatContext) -> bool {
+    if CallChainLayout::from_expression(AnyNodeRef::from(expr), context.source())
+        == CallChainLayout::Fluent
+    {
+        return false;
```

Comment:
@konstin - This feels wrong, but without it, we're dropping parentheses sometimes in call chain assignments. I assume the previous implementation was always returning `false` here, but I wonder if that was just a coincidence based on the definition of a call chain...

Reply:
Does black return false from the can omit function for these cases too?

Reply:
iirc `visitor.max_priority == OperatorPriority::Attribute && visitor.max_priority_count > 1` would return false, but i'm not sure if e.g. `(a).b().b()` is correctly handled here, i have to check tomorrow

Reply:
(I can check what Black returns here.)

Reply:
Real example that we no longer wrap properly:

```python
query = (
    UserMessage.select_for_update_query()
    .filter(
        user_profile=user_profile,
        message__recipient_id=stream_recipient_id,
    )
    .extra(
        where=[UserMessage.where_unread()],
    )
)
```

Reply:
Removing the `if` here fixed it:

```rust
// `[a, b].test[300].dot`
Expr::Attribute(ast::ExprAttribute {
    range: _,
    value,
    attr: _,
    ctx: _,
}) => {
    ...
    if has_parentheses(value, self.source) {
        self.update_max_priority(OperatorPriority::Attribute);
    }
    ...
}
```

Micha, do you know why that was necessarily? I went back to the original PR, removed the `if`, and tests passed.

Reply:
It could be related to the logic here

https://github.com/psf/black/blob/a44dc3d59eb46901f9fe893727280903df41fc20/src/black/brackets.py#L243-L250

The following method determines the max priority operator for black.

https://github.com/psf/black/blob/a44dc3d59eb46901f9fe893727280903df41fc20/src/black/brackets.py#L76-L134

which calls into the above method. 

The logic here is all reverse-engineered. There's a chance that I got it wrong or didn't notice that something is wrong because it created better formatting back then. 

I recommend changing this in isolation and then go through the created black differences manually.

Reply:
Removing this line on `main` doesn't cause any changes in the test fixtures.

I went through the exercise of differing `main` formatting of Zulip to the formatting you get if you remove this line -- here's the full diff:

```diff
diff --git a/zerver/lib/import_realm.py b/zerver/lib/import_realm.py
index c598959819..d93307fd91 100644
--- a/zerver/lib/import_realm.py
+++ b/zerver/lib/import_realm.py
@@ -1146,9 +1146,11 @@ def do_import_realm(import_dir: Path, subdomain: str, processes: int = 1) -> Rea
     # Ensure the invariant that there's always a realm-creation audit
     # log event, even if the export was generated by an export tool
     # that does not create RealmAuditLog events.
-    if not RealmAuditLog.objects.filter(
-        realm=realm, event_type=RealmAuditLog.REALM_CREATED
-    ).exists():
+    if (
+        not RealmAuditLog.objects.filter(
+            realm=realm, event_type=RealmAuditLog.REALM_CREATED
+        ).exists()
+    ):
         RealmAuditLog.objects.create(
             realm=realm,
             event_type=RealmAuditLog.REALM_CREATED,
diff --git a/zerver/lib/push_notifications.py b/zerver/lib/push_notifications.py
index 5e6e2e44ae..efba34096e 100644
--- a/zerver/lib/push_notifications.py
+++ b/zerver/lib/push_notifications.py
@@ -433,9 +433,11 @@ def send_android_push_notification(
             if reg_id == new_reg_id:
                 # I'm not sure if this should happen. In any case, not really actionable.
                 logger.warning("GCM: Got canonical ref but it already matches our ID %s!", reg_id)
-            elif not DeviceTokenClass.objects.filter(
-                token=new_reg_id, kind=DeviceTokenClass.GCM
-            ).count():
+            elif (
+                not DeviceTokenClass.objects.filter(
+                    token=new_reg_id, kind=DeviceTokenClass.GCM
+                ).count()
+            ):
                 # This case shouldn't happen; any time we get a canonical ref it should have been
                 # previously registered in our system.
                 #
diff --git a/zerver/management/commands/sync_ldap_user_data.py b/zerver/management/commands/sync_ldap_user_data.py
index a8a3bd259a..59683c9db3 100644
--- a/zerver/management/commands/sync_ldap_user_data.py
+++ b/zerver/management/commands/sync_ldap_user_data.py
@@ -42,12 +42,14 @@ def sync_ldap_user_data(
                     "Use the --force option if the mass deactivation is intended."
                 )
             for string_id in realms:
-                if not UserProfile.objects.filter(
-                    is_bot=False,
-                    is_active=True,
-                    realm__string_id=string_id,
-                    role=UserProfile.ROLE_REALM_OWNER,
-                ).exists():
+                if (
+                    not UserProfile.objects.filter(
+                        is_bot=False,
+                        is_active=True,
+                        realm__string_id=string_id,
+                        role=UserProfile.ROLE_REALM_OWNER,
+                    ).exists()
+                ):
                     raise Exception(
                         f"LDAP sync would have deactivated all owners of realm {string_id}. "
                         "This is most likely due "
diff --git a/zerver/tests/test_tornado.py b/zerver/tests/test_tornado.py
index 1475bd1046..1e4ef8870a 100644
--- a/zerver/tests/test_tornado.py
+++ b/zerver/tests/test_tornado.py
@@ -69,8 +69,10 @@ class TornadoWebTestCase(ZulipTestCase):
         if "HTTP_HOST" in kwargs:
             kwargs["headers"]["Host"] = kwargs["HTTP_HOST"]
             del kwargs["HTTP_HOST"]
-        return await self.http_client.fetch(
-            f"http://127.0.0.1:{self.port}{path}", method=method, **kwargs
+        return (
+            await self.http_client.fetch(
+                f"http://127.0.0.1:{self.port}{path}", method=method, **kwargs
+            )
         )
 
     def login_user(self, *args: Any, **kwargs: Any) -> None:
diff --git a/zerver/views/reactions.py b/zerver/views/reactions.py
index 316d82eb0d..b1f8317bcf 100644
--- a/zerver/views/reactions.py
+++ b/zerver/views/reactions.py
@@ -59,12 +59,14 @@ def remove_reaction(
         # corresponding code using the current data.
         emoji_code = get_emoji_data(message.sender.realm_id, emoji_name).emoji_code
 
-    if not Reaction.objects.filter(
-        user_profile=user_profile,
-        message=message,
-        emoji_code=emoji_code,
-        reaction_type=reaction_type,
-    ).exists():
+    if (
+        not Reaction.objects.filter(
+            user_profile=user_profile,
+            message=message,
+            emoji_code=emoji_code,
+            reaction_type=reaction_type,
+        ).exists()
+    ):
         raise ReactionDoesNotExistError
 
     # Unlike adding reactions, while deleting a reaction, we don't
diff --git a/zerver/views/realm_emoji.py b/zerver/views/realm_emoji.py
index 35bf25932c..2c5f22fe4d 100644
--- a/zerver/views/realm_emoji.py
+++ b/zerver/views/realm_emoji.py
@@ -55,9 +55,11 @@ def upload_emoji(
 
 
 def delete_emoji(request: HttpRequest, user_profile: UserProfile, emoji_name: str) -> HttpResponse:
-    if not RealmEmoji.objects.filter(
-        realm=user_profile.realm, name=emoji_name, deactivated=False
-    ).exists():
+    if (
+        not RealmEmoji.objects.filter(
+            realm=user_profile.realm, name=emoji_name, deactivated=False
+        ).exists()
+    ):
         raise ResourceNotFoundError(
             _("Emoji '{emoji_name}' does not exist").format(emoji_name=emoji_name)
         )
```

Reply:
Does the formatting now match Black's formatting? It otherwise mainly looks like a regression. I leave it to you to investigate this more.

Reply:
No, I thought it was clear that those do not match Black's formatting. I was reporting back for completeness.

Reply:
Ok, I think this is the relevant line in Black: https://github.com/psf/black/blob/a44dc3d59eb46901f9fe893727280903df41fc20/src/black/lines.py#L816

Reply:
Seems reasonable. Does that mean that we now return the same from `can_omit_parentheses` as black?

Reply:
Yeah